plugin/kubernetes/controller.go

```diff
@@ -491,63 +508,167 @@ func (dns *dnsControl) updateModifed() {
 	atomic.StoreInt64(&dns.modified, unix)
 }
 
-func (dns *dnsControl) Add(obj interface{})    { dns.updateModifed() }
-func (dns *dnsControl) Delete(obj interface{}) { dns.updateModifed() }
+func (dns *dnsControl) sendServiceUpdates(s *api.Service) {
+	var z []string
+	for i := range dns.zones {
+		name := ServiceFQDN(s, dns.zones[i])
+		if _, ok := dns.watched[name]; ok {
+			z = append(z, name)
+		}
+	}
+	if len(z) > 0 {
+		*dns.watchChan <- z
+	}
+}
+
+func (dns *dnsControl) sendPodUpdates(p *api.Pod) {
+	var z []string
+	for i := range dns.zones {
+		name := PodFQDN(p, dns.zones[i])
+		if _, ok := dns.watched[name]; ok {
+			z = append(z, name)
+		}
+	}
+	if len(z) > 0 {
+		*dns.watchChan <- z
+	}
+}
+
+func (dns *dnsControl) sendEndpointsUpdates(ep *api.Endpoints) {
+	var z []string
+	for _, zone := range dns.zones {
+		names := append(EndpointFQDN(ep, zone, dns.endpointNameMode), ServiceFQDN(ep, zone))
+		for _, name := range names {
+			if _, ok := dns.watched[name]; ok {
+				z = append(z, name)
+			}
+		}
+	}
+	if len(z) > 0 {
+		*dns.watchChan <- z
+	}
+}
+
+// xorSubsets returns an endpoints struct containing the Subsets that have changed (xor) between a and b
+func xorSubsets(a, b *api.Endpoints) *api.Endpoints {
```

Comment:
I am not sure I understand this function.

Reply:
It returns an endpoints struct containing the Subsets that have changed (xor) between a and b...

Reply:
We only want to notify client of things that changed...

Reply:
I understand what it is *for*. But it's a bit opaque to understand what it *does*. But looking at it in combination with the test helps.

Reply:
I'll expand the description...

Reply:
... and i renamed it to endpointsSubsetDiffs()